lib/OpenQA/WebAPI/Controller/Admin/Needle.pm

```diff
@@ -124,41 +124,70 @@ sub module {
 
 sub delete {
     my ($self) = @_;
-    my @removed_ids;
-    my @errors;
-    for my $needle_id (@{$self->every_param('id')}) {
-        my $needle = $self->db->resultset('Needles')->find($needle_id);
-        if (!$needle) {
-            push(
-                @errors,
-                {
-                    id           => $needle_id,
-                    display_name => $needle_id,
-                    message      => "Unable to find $needle_id"
-                });
-            next;
-        }
 
-        my $error = $needle->remove($self->current_user);
-        if ($error) {
-            push(
-                @errors,
-                {
-                    id           => $needle_id,
-                    display_name => $needle->filename,
-                    message      => $error
-                });
-        }
-        else {
-            push(@removed_ids, $needle_id);
-        }
+    # check whether Minion worker are available to get a nice error message instead of an inactive job
+    my $gru = $self->gru;
+    if (!$gru->has_workers) {
+        return $self->render(
+            json => {error => 'No Minion worker available. The <code>openqa-gru</code> service is likely not running.'}
+        );
     }
-    $self->emit_event('openqa_needle_delete', {id => \@removed_ids});
-    $self->render(
-        json => {
-            removed_ids => \@removed_ids,
-            errors      => \@errors
-        });
+
+    # enqueue Minion job delete needles with specified IDs
+    my %minion_args = (
+        needle_ids => $self->every_param('id'),
+        user_id    => $self->current_user->id,
+    );
+    my %minion_options = (
+        priority => 10,
+        ttl      => 60,
+    );
+    my $ids = $gru->enqueue(delete_needles => \%minion_args, \%minion_options);
+    my $minion_id;
+    if (ref $ids eq 'HASH') {
+        $minion_id = $ids->{minion_id};
+    }
+    my $minion     = $self->app->minion;
+    my $minion_job = $minion->job($minion_id);
+    if (!$minion_job) {
+        return $self->render(json => {error => 'Unable to enqueue Minion job for deleting needles.'});
+    }
```

Comment:
This condition should not be possible.

Reply:
I didn't know. I'll remove that check from the save needle task as well.

Reply:
The only case where Gru won't enqueue a job is if you set its `limit` option and it is exceeded. So it is safe to assume that it will always enqueue successfully.